### DISCUSSION TITLE
ZUGFeRD 2.2.0 Profil XRECHNUNG

### DISCUSSION THREAD

**Discussion by KlausV66** [stephanstapel/ZUGFeRD-csharp#307] [General]
Hello,
I'm looking for a component to support us to create the ZUGFeRD xml file and marriage it to the PDF A/3. 
Due to the fact that we are charging in B2B and B2G sector it's mandatory to use the profile xRECHNUNG based on the ZUGFeRD specification 2.2.0.
Does these functions are supported in your component? 
Thanks so much.
Klaus

**Comment by stephanstapel** [comment]
Hi Klaus,
I am certainly not dealing with marriages :)
I guess you are referring to embedding the xml file into the PDF A/3 file, right? In order to accomplish this functionalities, proper PDF libraries are what you are looking for. I have good experience with:
* https://www.pdf-tools.com/de/
* https://itextpdf.com/
* https://www.e-iceblue.com/Introduce/pdf-for-net-introduce.html?_gl=1*1njdgv2*_up*MQ..*_ga*MTM3Mzg0OTIzMC4xNzIxODM5MDA3*_ga_XLFT57F2BL*MTcyMTgzOTAwNi4xLjAuMTcyMTgzOTAwNi4wLjAuMA..*_ga_0P9F0Q9F33*MTcyMTgzOTAwNi4xLjAuMTcyMTgzOTAwNi4wLjAuMA..&gclid=CjwKCAjwzIK1BhAuEiwAHQmU3lmVbLLfQRMdCW1t6jGAHHqoFCQPTqjuZp5RhEvLuKBYK7LzlI_1zBoCBC8QAvD_BwE

**Reply by KlausV66** [reply]
Thanks Stephan of sharing the links, but not so cheap at all ;-)
Coming back to my question regarding the support of ZUGFeRD profile XRECHNUNG? Does it supported? 
Do you have a manual how to use it?

**Reply by stephanstapel** [reply]
Sorry, I can't relate to the statement. PDF tools are not cheap, that is right. Decent knowledge goes into developing these libraries and somebody has to pay these specialists.

Concerning your other question, please read the project page, it contains the requested information: https://github.com/stephanstapel/ZUGFeRD-csharp

**Comment by testtomat** [comment]
Hello,

here [https://to-works.de/wp/zaumzeug-c-zugferd-beispiel-implementierung-updated/](https://to-works.de/wp/zaumzeug-c-zugferd-beispiel-implementierung-updated/) you can find a c# sample project based on [https://www.mustangproject.org](https://www.mustangproject.org).

This project has a funtion to embed (marriage ;-) ) XML-Data in a PDF-document:

```
ZUGFeRDExporter ze = new ZUGFeRDExporter();
ze.PDFmakeA3compliant(pathToPdfDocumentIn, "Producer", "Creator", true);
ze.PDFattachZugferdFile(zugferdData);
ze.export(pathToPdfDocumentOut);    
```

**Reply by testtomat** [reply]
Mustangproject is Open Source - Permissive license (APL2, free for use in commercial and noncommercial applications)

**Reply by stephanstapel** [reply]
Good morning,
that's correct. Jochen tries to drop the code here:
https://github.com/ArtifexSoftware/Ghostscript.NET/pull/126

and I am keen on generalizing the code so that one can create a PDF A/3 with any xml (from any ZUGFeRD library). The code definitely needs refactoring anyway with a .ps template embedded in a c# code file currently.
Unfortunately, I cannot get Ghostscript.net to work:
https://github.com/ArtifexSoftware/Ghostscript.NET/issues/127

If you have experience with it, please let me know.

**Reply by testtomat** [reply]
Hi Stephan,
sorry, unfortunately I have no experience with Ghostscript.NET. I tested only embedding xml (from any ZUGFeRD library) using this mustangproject implementation. It works without any problems.

**Reply by stephanstapel** [reply]
Ah, I mixed two attempts here, sorry. The attempt to use PDFBox.net is an alternative way, compared to:
https://github.com/ArtifexSoftware/Ghostscript.NET/pull/126

ikvm.net comes with a huge overhead which imho is not acceptable for production use. Also the PDFBox version that was ported is really old, 1.8.x with version 4 currently in the works.

**Reply by hh-it-co** [reply]
Since i also had problems using Ghostscript.NET, like you have descripted here https://github.com/ArtifexSoftware/Ghostscript.NET/issues/127, i have implemented my own ghostscript-wrapper using Process-class, which works well.
Would be nice, to have Ghostscript.NET working, but maybe that may also help you for the time being.

**Reply by stephanstapel** [reply]
At least I'm not the only one...
Thanks for getting back to this here. Can you share the code so that it can get part of the library?

**Comment by hh-it-co** [comment]
`using System.Diagnostics;

namespace mynamespace
{
    public class GhostscriptWrapper
    {
        public string ExecuteGhostscript(string args)
        {
            try
            {
                // Create a new process
                Process process = new Process();

                // Configure process to use Ghostscript from PATH-variable or absolute path
                process.StartInfo.FileName = @"C:\Program Files\gs\gs10.03.1\bin\gswin64c.exe";
                // Pass the Ghostscript arguments
                process.StartInfo.Arguments = args;

                // Redirect output for capturing logs or errors
                process.StartInfo.RedirectStandardOutput = true;
                process.StartInfo.RedirectStandardError = true;
                process.StartInfo.UseShellExecute = false;
                process.StartInfo.CreateNoWindow = true;

                // Start the process
                process.Start();

                // Capture output and error streams
                string output = process.StandardOutput.ReadToEnd();
                string error = process.StandardError.ReadToEnd();

                // Wait for the process to finish
                process.WaitForExit();

                // Handle errors if any
                if (process.ExitCode != 0)
                {
                    throw new Exception($"Ghostscript execution failed: {error}");
                }

                // Return the output
                return output;
            }
            catch (Exception ex)
            {
                // Handle errors that might occur during process execution
                return $"Error during Ghostscript processing: {ex.Message}";
            }
        }
    }
}
`

Use it like that:

`
            // Build the argument string for Ghostscript
            string ghostscriptArgs = $"-sDEVICE=pdfwrite " +
                                     $"-dQUIET " +
...
                                     $"\"{inputPdfPath}\"";

            // Execute Ghostscript with the provided arguments
            string result = ghostscriptWrapper.ExecuteGhostscript(ghostscriptArgs);`

**Reply by stephanstapel** [reply]
Thanks! How do you use that to process pdf and xml file to create pdf a/3 with embedded xml?

**Reply by hh-it-co** [reply]
I followed this instructions https://www.ghostscript.com/blog/zugferd.html.
I got it working, but you have to consider license limitation of ghostscript.
